In [2]:
import pandas as pd
import warnings
warnings.filterwarnings('ignore')
df = pd.read_csv('Resources/client_dataset.csv')
df = df.drop_duplicates()
df = df.dropna()
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [3]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [4]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [5]:
# Use this space to do any additional research
# and familiarize yourself with the data.
#* -------------------------------------------
# df.info()
# df.head()
# df.sort_values(by='order_week', ascending=False).head()
# df.sort_values(by=['category']).head()
# count_jocabs = df[df[['first''last'] == 'Jacob','Mendoza'].value_counts()
# count_jocabs

"""
  if a specific client needed to be found in this data using their 
  fist and last names a new full_name column will need to be create where
  the first and last (name) columns will be joined
"""
df['full_name'] = df['first']+ ' ' + df['last']
df[df['full_name'] == 'Jacob Mendoza']['full_name'].value_counts()

"""
  in cases where you wish to display a single data point without an index it may 
  be appropriate to use .to_string(index=False). This will convert the data point
  to a string which can then be used seamlessly in a print statement
"""
df['full_name'].head(1).to_string(index=False)

'Donald Harding'

In [6]:
# What three item categories had the most entries?

"""
  in this case we need to count how many times each
  category is mentioned and return the top three most
  mentioned categories. a good method for handling this 
  is .value_count()

  @params subset, normalize, sort, ascending, dropna
  @returns the total times that a given value is
  counted in the data
  .value_count()
"""
categories = df[['category']].value_counts().head(3)
categories

category   
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [7]:
# For the category with the most entries, which subcategory had the most entries?

"""
  now that the most mentioned categories have been identified we need to know which
  of their subcategories is mentioned the most. one way to do this is to use .group_by().
  with this we can aggregate the data using the subcategories as our collections and
  then we just need to count their values.

  @params by, axis, level, as_index, sort, group_keys, observed, dropna
  @returns aggregated data collections
  .groupby()
"""
sub_cats = df[['category','subcategory']].groupby(['subcategory']).value_counts()
sub_cats.head(3)

subcategory        category   
accounting         software       3756
audio visual       electronics     732
bathroom supplies  consumables    6424
Name: count, dtype: int64

In [8]:
# Which five clients had the most entries in the data?
# Store the client ids of those top 5 clients in a list.
top_five = df[['full_name','client_id']].value_counts().head()
top_five

full_name        client_id
Jessica Reyes    33615        220
Angela Everett   66037        211
Bryan Myers      46820        209
Kendra Garrett   24741        207
Alexandra Young  38378        207
Name: count, dtype: int64

In [9]:
# How many total units (the qty column) did the client with the most entries order order?

"""
  There may be times that you only need a single value from a dataframe.
  in some cases it may be best to use .index to allow for finding the value
  using its placement index

  .index[0][0]
"""

top_client = df[['full_name', 'client_id']].value_counts().head(1)
top_client = pd.DataFrame(top_client)
name = top_client.index[0][0]
id = top_client.index[0][1]
total_units = df.loc[df['client_id'] == id, 'qty'].sum()
print(f'The client with the most entries is {name} with a total of {total_units} units bought! Good job {name}! Keep it up!')

The client with the most entries is Jessica Reyes with a total of 64313 units bought! Good job Jessica Reyes! Keep it up!


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [10]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty

"""
  There may be times when it is best to create a new column and store
  the aggregated data of one of more columns
"""
df['total_units_price'] = df['unit_price'] * df['qty']
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number           full_name  \
0     1096.80     762.71         7.50  105            1      Donald Harding   
1       24.95      15.09         1.49   21            0       Tiffany Myers   
2       13.52       7.86         1.68   39            6      Shannon Watson   
3       36.42      24.85         1.23   29            3        Nathan Baker   
4      195.10     108.17        46.43   20            1  Christina Schwartz   

   total_units_price  
0          115164.00  
1             523.95  
2             527.28  
3            1056.18  
4            3902.00

In [11]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.

"""
  We need to know how much it is going to cost to ship our units and we need to store that
  shipping price. Using an if-else statement should make it possible to add the correct value to the new shipping_price column
"""
if df['unit_weight'].all() >= 50:
  df['shipping_price'] = df['unit_weight'] * 10
else:
  df['shipping_price'] = df['unit_weight'] * 7


In [12]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%

"""
  In cases where you are working with floats it may be necessary to reformat pandas.
  To do this you can use pd.options.display.float_format = '{:.2f}'.format
"""
df['total_price'] = df['unit_price'] + df['shipping_price'] * .0925


In [13]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).

df['total_processing_cost'] = df.unit_cost + df.qty + df.shipping_price

In [14]:
# Create a column for the profit of each line using line cost and line price
df['profit'] = df.unit_price - df.unit_cost


## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [22]:
# Check your work using the totals above
"""
  For this we will use a lambda with an axis of 1
"""
verify = df.copy().head()
verify.apply(lambda row: print(f'Order ID {row.order_id} had a total price of ${float(row.total_units_price):,.2f}'), axis=1).dropna()

Order ID 8953482 had a total price of $115,164.00
Order ID 8069089 had a total price of $523.95
Order ID 1902144 had a total price of $527.28
Order ID 9031802 had a total price of $1,056.18
Order ID 1322274 had a total price of $3,902.00


Series([], dtype: object)

## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.


client_id	total_units

	58515	31158688

	37609	31158688

	57113	31158688

	46554	31158688

	92089	31158688


In [16]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
'''
  Find the total units for all clients, then return the top 5.
  First clean up and make a copy of the df, then groupby() can be used to
  group all the qty data points with by client_id. Then we can aggregate the
  qty(s) sum up their totals, sort the values and reset index
'''

top_five_totals = df.dropna(how='any').drop_duplicates().copy()
top_five_totals = top_five_totals.groupby('client_id').agg(
  unit_total = ('qty', 'sum')
  ).sort_values(by='unit_total', ascending=False).head(5).reset_index()
top_five_totals

client_id  unit_total
0      78965     4033302
1      86721     1010426
2      30143      846069
3      85931      644423
4      77354      506382

In [17]:
df.info()
top_five_totals.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 24 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   first                  54639 non-null  object 
 1   last                   54639 non-null  object 
 2   job                    54639 non-null  object 
 3   phone                  54639 non-null  object 
 4   email                  54639 non-null  object 
 5   client_id              54639 non-null  int64  
 6   order_id               54639 non-null  int64  
 7   order_date             54639 non-null  object 
 8   order_week             54639 non-null  int64  
 9   order_year             54639 non-null  int64  
 10  item_id                54639 non-null  object 
 11  category               54639 non-null  object 
 12  subcategory            54639 non-null  object 
 13  unit_price             54639 non-null  float64
 14  unit_cost              54639 non-null  float64
 15  un

In [18]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit.
'''
  For this problem we first need to make sure that we are comparing dataframes of the same length.
  The client_ids from the top_five_totals should be stored in a variable. With the top_five_ids the original
  dataframe can now be tested against to see if the client_ids from top_five_totals are in the original dataframe.
  If the condition check returns as true aggregate data from the original dataframe. The now aggregated data can be
  merged into the top_five_totals dataframe.
'''
# store top client_id(s)
top_five_ids = top_five_totals.client_id

'''
  Use .isin() to see if the client_id(s) from top_five_totals is in the original dataframe.
  If true then group data by id(s) and agg() datum.

  NOTE: agg() is used to aggregate datum related to a single column. In this case multiple columns
  need to be aggregated into one. One method to achieve this is to use lambda.
'''
# store aggregate data
agg_ttls = df[df['client_id'].isin(top_five_ids)].groupby('client_id').agg(
  total_shipping_price = ('shipping_price', 'sum'),
  total_revenue = ('unit_price', lambda x: (x * df.loc[x.index, 'qty']).sum()),
  total_profit = ('unit_price', lambda x: ((x - df.loc[x.index, 'unit_cost']) * df.loc[x.index, 'qty']).sum())
)

# merge agg_ttls
top_five_merged = top_five_totals.merge(agg_ttls, on='client_id').reindex()

# display top_five_totals
top_five_merged.style

In [19]:

# Format the data and rename the columns to names suitable for presentation.
top_five_merged = top_five_merged.rename(columns={
  'client_id': 'Client ID', 
  'unit_total':	'Total Units Ordered', 
  'total_shipping_price':	'Total Spent on Shipping', 
  'total_revenue': 'Total Revenue Earned', 
  'total_profit': 'Total Profits Earned'})

# Define a function that converts a dollar amount to millions.
def amilli(x):
  return x / 1_000_000

# Rename the columns to reflect the change in the money format.
# Define the money columns.
top_five_merged['Total Spent on Shipping'] = top_five_merged['Total Spent on Shipping'].map('${:,.2f}'.format)
top_five_merged['Total Profit (millions)'] = amilli(top_five_merged['Total Profits Earned']).map('${:.2f}M'.format)
top_five_merged['Total Revenue (millions)'] = amilli(top_five_merged['Total Revenue Earned']).map('${:.2f}M'.format)
top_five_merged['Total Profits Earned'] = top_five_merged['Total Profits Earned'].map('${:,.2f}'.format)
top_five_merged['Total Revenue Earned'] = top_five_merged['Total Revenue Earned'].map('${:,.2f}'.format)

# Apply the currency_format_millions function to only the money columns. 
top_five_merged

Client ID  Total Units Ordered Total Spent on Shipping  \
0      78965              4033302               $5,168.31   
1      86721              1010426               $4,308.22   
2      30143               846069               $1,510.11   
3      85931               644423               $1,934.10   
4      77354               506382               $1,795.01   

  Total Revenue Earned Total Profits Earned Total Profit (millions)  \
0      $595,047,041.80      $191,816,971.09                $191.82M   
1      $101,583,079.67       $27,627,205.81                 $27.63M   
2       $21,525,051.13        $8,428,492.43                  $8.43M   
3      $207,192,783.92       $89,192,460.51                 $89.19M   
4       $42,036,711.65        $6,341,005.84                  $6.34M   

  Total Revenue (millions)  
0                 $595.05M  
1                 $101.58M  
2                  $21.53M  
3                 $207.19M  
4                  $42.04M

In [20]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
top_five_merged['Total Profit (millions)'].sort_values()


0    $191.82M
1     $27.63M
4      $6.34M
2      $8.43M
3     $89.19M
Name: Total Profit (millions), dtype: object